### PR TITLE
Increased the performance of blockchain statistics plugin

### PR DESCRIPTION
**Pull request by AKorpusenko**
1. Now correct statistics sends right after running `golosd` and getting second block (minimum two blocks are needed to calc statistics delta).
2. Statistics sends in separated thread what increases plugin's performance.

### CODE REVIEW

File: libraries/plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+    share_type sbd_to_be_converted = 0;                     ///< Amount of SBD to be converted
+    uint32_t sbd_conversion_requests_filled = 0;          ///< SBD conversion requests filled
+    share_type steem_converted = 0;                         ///< Amount of STEEM that was converted
+    uint32_t limit_orders_created = 0;                    ///< Limit orders created
+    uint32_t limit_orders_filled = 0;                     ///< Limit orders filled
+    uint32_t limit_orders_cancelled = 0;                  ///< Limit orders cancelled
+    uint32_t total_pow = 0;                               ///< POW submitted
+    uint128_t estimated_hashpower = 0;                     ///< Estimated average hashpower over interval
+
+    std::vector < std::string > get_as_string () {
```
Comment (kotbegemot) [inline libraries/plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: move to cpp

File: libraries/plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+        result.push_back("steem_converted:" + std::string(steem_converted));
+        result.push_back("limit_orders_created:" + std::to_string(limit_orders_created));
+        result.push_back("limit_orders_filled:" + std::to_string(limit_orders_filled));
+        result.push_back("limit_orders_cancelled:" + std::to_string(limit_orders_cancelled));
+        result.push_back("total_pow:" + std::to_string(total_pow));
+        result.push_back("estimated_hashpower:" + std::string(estimated_hashpower));
+        return result;
+    }
+
+    std::vector < std::string > calculate_delta_with (const bucket_object & b) {
```
Comment (kotbegemot) [inline libraries/plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: move to cpp

File: libraries/plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        if (b.total_pow - total_pow != 0 ) {
+            result.push_back("total_pow:" + std::to_string(b.total_pow - total_pow) + "|c");
+        }
+        if (b.estimated_hashpower - estimated_hashpower != 0 ) {
+            result.push_back("estimated_hashpower:" + std::string(b.estimated_hashpower - estimated_hashpower) + "|c");
+        }
+        return result;
+    }
+
+    void operator=(const bucket_object & b) {
```
Comment (kotbegemot) [inline libraries/plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: move to cpp

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+    share_type sbd_to_be_converted = 0;                     ///< Amount of SBD to be converted
+    uint32_t sbd_conversion_requests_filled = 0;          ///< SBD conversion requests filled
+    share_type steem_converted = 0;                         ///< Amount of STEEM that was converted
+    uint32_t limit_orders_created = 0;                    ///< Limit orders created
+    uint32_t limit_orders_filled = 0;                     ///< Limit orders filled
+    uint32_t limit_orders_cancelled = 0;                  ///< Limit orders cancelled
+    uint32_t total_pow = 0;                               ///< POW submitted
+    uint128_t estimated_hashpower = 0;                     ///< Estimated average hashpower over interval
+
+    std::vector < std::string > get_as_string () {
```
Comment (kotbegemot) [inline plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: move to cpp

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+        result.push_back("steem_converted:" + std::string(steem_converted));
+        result.push_back("limit_orders_created:" + std::to_string(limit_orders_created));
+        result.push_back("limit_orders_filled:" + std::to_string(limit_orders_filled));
+        result.push_back("limit_orders_cancelled:" + std::to_string(limit_orders_cancelled));
+        result.push_back("total_pow:" + std::to_string(total_pow));
+        result.push_back("estimated_hashpower:" + std::string(estimated_hashpower));
+        return result;
+    }
+
+    std::vector < std::string > calculate_delta_with (const bucket_object & b) {
```
Comment (kotbegemot) [inline plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: move to cpp

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        if (b.total_pow - total_pow != 0 ) {
+            result.push_back("total_pow:" + std::to_string(b.total_pow - total_pow) + "|c");
+        }
+        if (b.estimated_hashpower - estimated_hashpower != 0 ) {
+            result.push_back("estimated_hashpower:" + std::string(b.estimated_hashpower - estimated_hashpower) + "|c");
+        }
+        return result;
+    }
+
+    void operator=(const bucket_object & b) {
```
Comment (kotbegemot) [inline plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: move to cpp

Comment (AKorpusenko) [reply plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: Fixed

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp (hunk 5 above)
Comment (AKorpusenko) [reply plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: Fixed

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp (hunk 4 above)
Comment (AKorpusenko) [reply plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: Fixed

File: libraries/plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp (hunk 3 above)
Comment (AKorpusenko) [reply libraries/plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: Fixed

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+namespace golos {
+namespace plugins {
+namespace blockchain_statistics {
+
+using namespace golos::chain;
+
+struct runtime_bucket_object {
+
+    std::vector < std::string > get_as_string ();
```
Comment (kotbegemot) [inline plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: move to cpp

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    std::vector < std::string > calculate_delta_with (const bucket_object & b);
```
Comment (kotbegemot) [inline plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: move to cpp

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp (hunk 7 above)
Comment (AKorpusenko) [reply plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: What .cpp file do you mean? I don't understand at all

File: plugins/blockchain_statistics/statistics_sender.cpp [hunk 9]
```diff
@@ -15,22 +15,32 @@
...
 }
 
 bool statistics_sender::can_start() {
     return !recipient_endpoint_set.empty();
 }
 
 void statistics_sender::push(const std::string & str) {
-    boost::asio::io_service io_service;
+    ios.post ([this, str]() {
+        boost::asio::ip::udp::socket socket(this->ios, boost::asio::ip::udp::endpoint(boost::asio::ip::udp::v4(), 0));
```
Comment (afalaleev) [inline plugins/blockchain_statistics/statistics_sender.cpp]: You shouldn't create this socket on each push.
Move this socket to statistics_sender.

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/plugin.hpp [hunk 10]
```diff
@@ -8,6 +8,7 @@
 #include <boost/program_options.hpp>
 
 #include <golos/plugins/blockchain_statistics/bucket_object.hpp>
```
Comment (afalaleev) [inline plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/plugin.hpp]: Are you really need this include in plugin.hpp?

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp (hunk 7 above)
Comment (afalaleev) [reply plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: Alexander asked you to move this structure to plugin.cpp.

I think, you may keep runtime_bucket_object in separated header.

File: plugins/blockchain_statistics/plugin.cpp [hunk 11]
```diff
@@ -18,11 +20,6 @@ namespace golos {
...
 namespace blockchain_statistics {
 
-std::string increment_counter(std::string name);
-std::string increment_counter(std::string name, uint32_t value);
-std::string increment_counter(std::string name, fc::uint128_t value);
-std::string increment_counter(std::string name, share_type value);
-
 using namespace golos::protocol;
 
 struct plugin::plugin_impl final {
```
Comment (afalaleev) [inline plugins/blockchain_statistics/plugin.cpp]: Are you really need to track buckets for 1 minute, 1 hour and so on?
Your statistics_sender sends information only about last changes. 
You need to have previous values for operations, and don't need to have information about time periods.

Simplify your code: remove statistics from shared memory and store only runtime information with previous counters.

Comment (AKorpusenko) [reply plugins/blockchain_statistics/plugin.cpp]: @afalaleev , @kotbegemot , this heritage of old code was intended to know statistics per minute, hour, day, week or month. As we do not have API for `blockchain_statistics` plugin there are two ways: 
1. Add API and expand the functionality
2. Remove all shared memory buckets, but it causes some troubles and makes impossible to get absolutely correct statistics

File: plugins/blockchain_statistics/plugin.cpp [hunk 12]
```diff
@@ -568,44 +342,239 @@ void plugin::plugin_startup() {
...
-
-std::string increment_counter(std::string name, uint32_t value) {
-    std::string res = name + ":";
-    std::string num = std::to_string(value);
-    res += num + "|c";
-
-    return res;
+    if (b.blocks - blocks != 0 ) {
+        result.push_back("blocks:" + std::to_string(b.blocks - blocks) + "|c");
+    }
```
Comment (afalaleev) [inline plugins/blockchain_statistics/plugin.cpp]: Return the increment_counter() - it's more readable, than now:
```
...
increment_counter(result, "blocks", b.blocks - blocks);
increment_counter(result, "bandwidth", b.bandwidth - bandwidth);
...
```
Where increment_function() is:
```
void increment_counter(std::vector < std::string > &result, std::string name, uint32_t value) {
    if (value != 0) {
         result.push_back(name + std::to_string(value) + "|c");
    }
}
```

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp [hunk 13]
```diff
@@ -12,25 +10,12 @@ namespace blockchain_statistics {
...
-
-struct bucket_object
-        : public object<bucket_object_type, bucket_object> {
-    template<typename Constructor, typename Allocator>
-    bucket_object(Constructor &&c, allocator<Allocator> a) {
-        c(*this);
-    }
-
-    id_type id;
+    std::vector < std::string > get_as_string ();
```
Comment (kotbegemot) [inline plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: move to implementation

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp [hunk 14]
```diff
@@ -12,25 +10,12 @@ namespace blockchain_statistics {
...
+    std::vector < std::string > calculate_delta_with (const runtime_bucket_object & b);
```
Comment (kotbegemot) [inline plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: move to implementation

**Comment by kotbegemot** [comment]
@AKorpusenko 
blockchain statistics plugin to change the name to a state plugin.

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp (hunk 14 above)
Comment (AKorpusenko) [reply plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: Done

File: plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp (hunk 13 above)
Comment (AKorpusenko) [reply plugins/blockchain_statistics/include/golos/plugins/blockchain_statistics/runtime_bucket_object.hpp]: Done

File: plugins/blockchain_statistics/plugin.cpp (hunk 12 above)
Comment (AKorpusenko) [reply plugins/blockchain_statistics/plugin.cpp]: Done

**Comment by AKorpusenko** [comment]
@kotbegemot , are you sure that the name `state` instead of `blockchain_statistics` is ok?

**Comment by kotbegemot** [comment]
@AKorpusenko statd

**Comment by AKorpusenko** [comment]
@kotbegemot , done. Now `blockchain_statistics` is `statsd`.
@kotbegemot and @afalaleev , please review the code. I hope it's ok now and branch is ready to merge.

File: plugins/statsd/CMakeLists.txt [hunk 15]
```diff
@@ -12,12 +12,12 @@ list(APPEND CURRENT_TARGET_SOURCES
 )
 
 if(BUILD_SHARED_LIBRARIES)
-    add_library(golos_blockchain_statistics SHARED
+    add_library(golos_statsd SHARED
```
Comment (kotbegemot) [inline plugins/statsd/CMakeLists.txt]: Use Target CURRENT_TARGET

File: plugins/statsd/CMakeLists.txt [hunk 16]
```diff
@@ -12,12 +12,12 @@ list(APPEND CURRENT_TARGET_SOURCES
...
         ${CURRENT_TARGET_HEADERS}
         ${CURRENT_TARGET_SOURCES}
     )
 else()
-    add_library(golos_blockchain_statistics STATIC
+    add_library(golos_statsd STATIC
```
Comment (kotbegemot) [inline plugins/statsd/CMakeLists.txt]: Use Target CURRENT_TARGET

File: plugins/statsd/CMakeLists.txt (hunk 15 above)
Comment (AKorpusenko) [reply plugins/statsd/CMakeLists.txt]: Done

File: plugins/statsd/CMakeLists.txt (hunk 16 above)
Comment (AKorpusenko) [reply plugins/statsd/CMakeLists.txt]: Done

**Pull request merged**